src/metabase/sync/sync_metadata/fields.clj

```diff
@@ -39,29 +39,62 @@
   * In general the methods in these namespaces return the number of rows updated; these numbers are summed and used
     for logging purposes by higher-level sync logic."
   (:require
+   [metabase.driver :as driver]
+   [metabase.driver.util :as driver.u]
    [metabase.models.table :as table]
+   [metabase.sync.fetch-metadata :as fetch-metadata]
    [metabase.sync.interface :as i]
-   [metabase.sync.sync-metadata.fields.fetch-metadata :as fetch-metadata]
+   [metabase.sync.sync-metadata.fields.our-metadata :as our-metadata]
    [metabase.sync.sync-metadata.fields.sync-instances :as sync-instances]
    [metabase.sync.sync-metadata.fields.sync-metadata :as sync-metadata]
    [metabase.sync.util :as sync-util]
+   [metabase.util.log :as log]
    [metabase.util.malli :as mu]
-   [metabase.util.malli.schema :as ms]))
+   [metabase.util.malli.schema :as ms]
+   [toucan2.core :as t2]))
 
 ;;; +----------------------------------------------------------------------------------------------------------------+
 ;;; |                                            PUTTING IT ALL TOGETHER                                             |
 ;;; +----------------------------------------------------------------------------------------------------------------+
 
-(mu/defn ^:private sync-and-update! :- ms/IntGreaterThanOrEqualToZero
-  "Sync Field instances (i.e., rows in the Field table in the Metabase application DB) for a Table, and update metadata
+ (mu/defn ^:private sync-and-update! :- ms/IntGreaterThanOrEqualToZero
+   "Sync Field instances (i.e., rows in the Field table in the Metabase application DB) for a Table, and update metadata
   properties (e.g. base type and comment/remark) as needed. Returns number of Fields synced."
-  [table       :- i/TableInstance
-   db-metadata :- [:set i/TableMetadataField]]
-  (+ (sync-instances/sync-instances! table db-metadata (fetch-metadata/our-metadata table))
+   [table       :- i/TableInstance
+    db-metadata :- [:set i/TableMetadataField]]
+   (+ (sync-instances/sync-instances! table db-metadata (our-metadata/our-metadata table))
      ;; Now that tables are synced and fields created as needed make sure field properties are in sync.
      ;; Re-fetch our metadata because there might be somethings that have changed after calling
      ;; `sync-instances`
-     (sync-metadata/update-metadata! table db-metadata (fetch-metadata/our-metadata table))))
+      (sync-metadata/update-metadata! table db-metadata (our-metadata/our-metadata table))))
+
+(mu/defn sync-fields-for-db!
+  "Sync the Fields in the Metabase application database for a specific `table`."
+  [database :- i/DatabaseInstance]
```

Comment:
you can make this take `tables` and create a `[schema name] -> table` mapping that you can use inside the reducer to get the table.

Always good to reduce the number of DB calls

Reply:
I was also bothered by the N+1 table queries, but ended up not commenting as for large N we'd rather not fetch all the tables into memory.

It seems unfortunate to me that there's a deep nest of methods to require a full table model, because in practice they're mostly using the id, name, and schema mostly. I suspect we could just add one or two more table fields to the educed rows and get rid of this query, but it would be brittle to use a partially populated instance, and refactoring all those methods to depend only on their required table fields would be pretty hairy and ugly.

Reply:
What Chris said, I'm fetching each table because I'm avoiding fetching all the tables into memory. The good thing is app DB queries should be quite cheap compared to customer DB queries, so this is still a huge win.

In the future we can avoid the N+1 by doing a refactoring like the one I did for `sync-fks`  as I did here https://github.com/metabase/metabase/pull/39679/files#diff-e09c063342055336a262f66623a6a1783a89871e4426c33c92f8f9b5f2bcaaec

But I don't think it's worth it. That refactoring took longer than I thought because of the quirks between different app DBs.

Reply:
okay, that makes sense; in that case, you need to "not" fetch all tables in metabase.sync.sync-metadata.fields line 117